### ISSUE TITLE
Making the vis.DataSet allow to receive an Object representing a collection as input

### ISSUE THREAD

**Issue by felixhayashi** [almende/vis#583]
Hi,

it would be great if the DataSet could be enhanced to receive an Object of the structure `Object<id, Element>` when calling `update()`, `remove()` or `add()` instead of just an Array or an Object representing a single element (node/edge).

The point is, often it is more handy or necessary to deal with "hashmaps" that have ids as keys and elements as values and I always have to convert this collection to an array eventually, in order to manipulate the DataSet.

I guess the only reason this feature doesn't exist is because you cannot differentiate between whether an object is supposed to be a single entry in the dataset or if it represents a collection when passing it as parameter?

-Felix

**Comment by josdejong** [comment]
Yes makes sense to be able to add an Object containing a hashmap with items. 

You are right, we need a way there to distinguish between a single item Object or an Object holding a hashmap with items. I guess we could do that by either:
-  passing an extra parameter like `add(myObjects, {type: 'hashmap'})`, where type can be `'auto'` (default), `'single'`, `'array'`, or `'hashmap'` (or different naming).
- creating separate methods like `addHashMap(myObjects)`

Alternatively, we could deprecate support for adding individual items. That will require the cases where individual items are added/updated to be packed in an array, like `add([newItem])` instead of `add(newItem)`. This will be backwards incompatible, but there is at least a relatively easy way to throw errors when still called as `add(newItem)`: we expect all properties to hold an Object, which will not be the case then.

**Comment by felixhayashi** [comment]
Hi @josdejong,

thanks for picking this up. :) 

Many good ideas! I think the second way suggested is good for being backwards compatible, then e.g. for the `add` operation you need `addHashMap`, `addItem` and `addArray` the old `add` could still be used as convenience-proxy that only accepts an object (item) or an array. Then you would be backwards compatible and at the same time it is easy and straight forward to use these new dataset methods.

The first suggestion is fine too but has the overhead that a user often has to specify what data type is inserted/updated/added. This might be error prone as it is easily overlooked. If you force the second "type"-param then it is not backwards compatible anymore.

So in my opinion, your second approach is more robust.

Regards
Felix

**Comment by josdejong** [comment]
Both first two ideas are backwards compatible, I forgot to mention that this second parameter `{type: ...}` should be optional and when omitted, results in the current behavior. Similar to how the method `get` works right now, where you can specify a `{returnType: 'Array' | 'Object' | 'DataTable'}` (see [docs](http://visjs.org/docs/dataset.html#Data_Selection)). 

If we go for the second solution, we should also introduce methods like `getItem`, `getArray`, `getHashMap`, to be consistent with the `add*`, `update*` and `remove*` methods. This will bloat the API. But on the other hand, the methods are self explaining and easy to understand.

**Comment by felixhayashi** [comment]
Yes, it would bloat the api a little, this is the negative side of this approach. Yet, the concept is easy to understand (I am used to it from Java where, due to Java's strict typing, you need to define explicit setters and getters if you deal with different input/output types).

On the other hand, the advantage of the first approach is to offer the user a central method for input/output, however, if the second options-param is optional, then people might overlook it and pass a hashmap which regarded as single item by the system and is a source for errors. Of course, it could be argued, this could also happen with the second approach, but the first approach explicitly mentions that it is possible to pass a hashmap to e.g. `add`. 

Still I think you are right and it is better to go with the first suggestion, I mean it also somewhat complements the `returnType` option, as you said and it binds similar operations to one method, probably allowing it to be documented better.

So at the moment I favour the first approach again. Still, both have their pros/cons :)
